File: backend/app/services/service_order_beneficiary.py
"""Service order beneficiary + execution helpers (product model, no new entities)."""

from __future__ import annotations

from typing import Any, Literal, Optional, Tuple

BeneficiaryKind = Literal["client", "candidate", "employee"]

EXECUTION_INLINE = "inline"
EXECUTION_HANDOFF = "handoff"


def resolve_beneficiary(
    *,
    company_id: Optional[str],
    candidate_id: Optional[str],
    employee_id: Optional[str],
) -> Tuple[Optional[BeneficiaryKind], Optional[str]]:
    """Return (beneficiary_kind, beneficiary_id) for a service order."""
    cid = str(company_id or "").strip() or None
    cand = str(candidate_id or "").strip() or None
    emp = str(employee_id or "").strip() or None
    owners = [
        ("client", cid),
        ("candidate", cand),
        ("employee", emp),
    ]
    active = [(kind, oid) for kind, oid in owners if oid]
    if len(active) != 1:
        return None, None
    return active[0][0], active[0][1]  # type: ignore[return-value]


def beneficiary_from_order(order: Any) -> Tuple[Optional[BeneficiaryKind], Optional[str]]:
    return resolve_beneficiary(
        company_id=getattr(order, "company_id", None),
        candidate_id=getattr(order, "candidate_id", None),
        employee_id=getattr(order, "employee_id", None),
    )


# --- Customer (Bill-To) + per-line Beneficiary (Variant B) -----------------
CustomerKind = BeneficiaryKind
_VALID_ROLE_KINDS = ("client", "candidate", "employee")


def _normalize_role_kind(value: Any) -> Optional[BeneficiaryKind]:
    kind = str(value or "").strip().lower()
    return kind if kind in _VALID_ROLE_KINDS else None  # type: ignore[return-value]


def resolve_customer(order: Any) -> Tuple[Optional[CustomerKind], Optional[str]]:
    """Who pays (Bill-To). Canonical customer_kind/customer_id, else typed owner."""
    kind = _normalize_role_kind(getattr(order, "customer_kind", None))
    cid = str(getattr(order, "customer_id", None) or "").strip() or None
    if kind and cid:
        return kind, cid
    # Fallback for legacy orders (pre-migration / typed owner materialization).
    return beneficiary_from_order(order)


def order_customer_columns(
    customer_kind: Any,
    customer_id: Any,
) -> dict[str, Optional[str]]:
    """Materialize customer into the typed FK columns used by joins/scope/billing."""
    kind = _normalize_role_kind(customer_kind)
    cid = str(customer_id or "").strip() or None
    cols: dict[str, Optional[str]] = {"company_id": None, "candidate_id": None, "employee_id": None}
    if not kind or not cid:
        return cols
    if kind == "client":
        cols["company_id"] = cid
    elif kind == "candidate":
        cols["candidate_id"] = cid
    elif kind == "employee":
        cols["employee_id"] = cid
    return cols


def resolve_item_beneficiary(
    item: Any,
    order: Any = None,
) -> Tuple[Optional[BeneficiaryKind], Optional[str]]:
    """Who receives this line. Item beneficiary, else the order customer."""
    kind = _normalize_role_kind(getattr(item, "beneficiary_kind", None))
    bid = str(getattr(item, "beneficiary_id", None) or "").strip() or None
    if kind and bid:
        return kind, bid
    if order is not None:
        return resolve_customer(order)
    return None, None


def context_vacancy_id(order: Any) -> Optional[str]:
    vid = str(getattr(order, "vacancy_id", None) or "").strip()
    return vid or None


def execution_meta(meta: Any) -> dict[str, Any]:
    if not isinstance(meta, dict):
        return {}
    raw = meta.get("execution")
    if isinstance(raw, dict):
        return dict(raw)
    mode = str(meta.get("execution_mode") or "").strip().lower()
    if mode in (EXECUTION_INLINE, EXECUTION_HANDOFF):
        return {"mode": mode, "handoff_action": meta.get("handoff_action")}
    return {}


def service_execution_mode(meta: Any) -> str:
    block = execution_meta(meta)
    mode = str(block.get("mode") or EXECUTION_INLINE).strip().lower()
    return EXECUTION_HANDOFF if mode == EXECUTION_HANDOFF else EXECUTION_INLINE


def service_handoff_action(meta: Any) -> Optional[str]:
    block = execution_meta(meta)
    action = str(block.get("handoff_action") or meta.get("handoff_action") if isinstance(meta, dict) else "").strip()
    return action or None


def service_requires_handoff(meta: Any) -> bool:
    return service_execution_mode(meta) == EXECUTION_HANDOFF and bool(service_handoff_action(meta))


def build_execution_meta(*, mode: str, handoff_action: Optional[str] = None) -> dict[str, Any]:
    normalized = EXECUTION_HANDOFF if str(mode or "").strip().lower() == EXECUTION_HANDOFF else EXECUTION_INLINE
    block: dict[str, Any] = {"mode": normalized}
    if normalized == EXECUTION_HANDOFF and handoff_action:
        block["handoff_action"] = str(handoff_action).strip()
    return {"execution": block, "execution_mode": normalized}


def stamp_item_execution_meta(item_meta: Optional[dict[str, Any]], service_meta: Any) -> dict[str, Any]:
    """Snapshot catalog execution onto line item (stable after catalog edits)."""
    out = dict(item_meta or {})
    if isinstance(out.get("execution"), dict):
        return out
    block = execution_meta(service_meta)
    if not block:
        block = {"mode": EXECUTION_INLINE}
    out["execution"] = block
    out["execution_mode"] = block.get("mode") or EXECUTION_INLINE
    if block.get("handoff_action"):
        out["handoff_action"] = block.get("handoff_action")
    return out


def item_execution_mode(item_meta: Any, service_meta: Any = None) -> str:
    return service_execution_mode(item_meta if execution_meta(item_meta) else service_meta)


def item_handoff_action(item_meta: Any, service_meta: Any = None) -> Optional[str]:
    if execution_meta(item_meta):
        return service_handoff_action(item_meta)
    return service_handoff_action(service_meta)


def item_requires_handoff(item_meta: Any, service_meta: Any = None) -> bool:
    mode = item_execution_mode(item_meta, service_meta)
    if mode != EXECUTION_HANDOFF:
        return False
    return bool(item_handoff_action(item_meta, service_meta))

File: hostflow-frontend/src/modules/services/serviceOrderBeneficiary.ts
/** Service order beneficiary + catalog execution helpers.
 *
 * Params use structural shapes on purpose so both `api/types` and
 * `api/types/service` order/service variants are accepted without casts.
 */

export type ServiceBeneficiaryKind = 'client' | 'candidate' | 'employee'

export type ServiceExecutionMode = 'inline' | 'handoff'

type OrderRoleShape = {
  beneficiary_kind?: ServiceBeneficiaryKind | string | null
  company_id?: string | null
  client_id?: string | null
  candidate_id?: string | null
  employee_id?: string | null
  customer_kind?: ServiceBeneficiaryKind | string | null
  customer_id?: string | null
}

type CatalogMetaShape = { meta?: Record<string, unknown> | null } | null | undefined

type LineExecutionShape = {
  execution_mode?: string | null
  handoff_action?: string | null
  meta?: Record<string, unknown> | null
  service?: { meta?: Record<string, unknown> | null } | null
}

function normalizeKind(value: unknown): ServiceBeneficiaryKind | null {
  const kind = String(value || '').trim().toLowerCase()
  return kind === 'client' || kind === 'candidate' || kind === 'employee' ? kind : null
}

export function orderBeneficiaryKind(order: OrderRoleShape): ServiceBeneficiaryKind | null {
  const explicit = normalizeKind(order.beneficiary_kind)
  if (explicit) return explicit
  if (order.company_id || order.client_id) return 'client'
  if (order.candidate_id) return 'candidate'
  if (order.employee_id) return 'employee'
  return null
}

export function orderBeneficiaryId(order: OrderRoleShape): string | null {
  const kind = orderBeneficiaryKind(order)
  if (kind === 'client') return String(order.company_id || order.client_id || '').trim() || null
  if (kind === 'candidate') return String(order.candidate_id || '').trim() || null
  if (kind === 'employee') return String(order.employee_id || '').trim() || null
  return null
}

/** Bill-To / who pays. Canonical customer_*, else falls back to the typed owner. */
export function orderCustomerKind(order: OrderRoleShape): ServiceBeneficiaryKind | null {
  const explicit = normalizeKind(order.customer_kind)
  if (explicit) return explicit
  return orderBeneficiaryKind(order)
}

export function orderCustomerId(order: OrderRoleShape): string | null {
  if (order.customer_id) return String(order.customer_id).trim() || null
  return orderBeneficiaryId(order)
}

/** Who receives a given line. Item beneficiary, else the order customer. */
export function itemBeneficiaryKind(
  item: { beneficiary_kind?: ServiceBeneficiaryKind | string | null },
  order?: OrderRoleShape | null,
): ServiceBeneficiaryKind | null {
  const explicit = normalizeKind(item.beneficiary_kind)
  if (explicit) return explicit
  return order ? orderCustomerKind(order) : null
}

export function itemBeneficiaryId(
  item: { beneficiary_kind?: ServiceBeneficiaryKind | string | null; beneficiary_id?: string | null },
  order?: OrderRoleShape | null,
): string | null {
  if (item.beneficiary_id) return String(item.beneficiary_id).trim() || null
  if (normalizeKind(item.beneficiary_kind)) return null
  return order ? orderCustomerId(order) : null
}

export function catalogExecutionMode(service: CatalogMetaShape): ServiceExecutionMode {
  const meta = service?.meta
  if (!meta || typeof meta !== 'object') return 'inline'
  const block = (meta as Record<string, unknown>).execution
  if (block && typeof block === 'object') {
    const mode = String((block as Record<string, unknown>).mode || '').trim().toLowerCase()
    return mode === 'handoff' ? 'handoff' : 'inline'
  }
  const mode = String((meta as Record<string, unknown>).execution_mode || '').trim().toLowerCase()
  return mode === 'handoff' ? 'handoff' : 'inline'
}

export function catalogHandoffAction(service: CatalogMetaShape): string | null {
  const meta = service?.meta
  if (!meta || typeof meta !== 'object') return null
  const block = (meta as Record<string, unknown>).execution
  if (block && typeof block === 'object') {
    const action = String((block as Record<string, unknown>).handoff_action || '').trim()
    if (action) return action
  }
  const action = String((meta as Record<string, unknown>).handoff_action || '').trim()
  return action || null
}

export function itemExecutionMode(item: LineExecutionShape): ServiceExecutionMode {
  const fromApi = String(item.execution_mode || '').trim().toLowerCase()
  if (fromApi === 'handoff' || fromApi === 'inline') return fromApi
  const meta = item.meta
  if (meta && typeof meta === 'object' && meta.execution && typeof meta.execution === 'object') {
    const mode = String((meta.execution as Record<string, unknown>).mode || '').trim().toLowerCase()
    if (mode === 'handoff') return 'handoff'
  }
  return catalogExecutionMode(item.service ?? null)
}

export function itemHandoffAction(item: LineExecutionShape): string | null {
  const fromApi = String(item.handoff_action || '').trim()
  if (fromApi) return fromApi
  const meta = item.meta
  if (meta && typeof meta === 'object' && meta.execution && typeof meta.execution === 'object') {
    const action = String((meta.execution as Record<string, unknown>).handoff_action || '').trim()
    if (action) return action
  }
  return catalogHandoffAction(item.service ?? null)
}

export function beneficiaryKindLabel(
  kind: ServiceBeneficiaryKind | null | undefined,
  t: (key: string, options?: { defaultValue?: string }) => string,
): string {
  if (!kind) return t('app.services.beneficiary.unknown', { defaultValue: 'Not specified' })
  return t(`app.services.beneficiary.${kind}`, {
    defaultValue: kind === 'client' ? 'Client' : kind === 'candidate' ? 'Candidate' : 'Employee',
  })
}
